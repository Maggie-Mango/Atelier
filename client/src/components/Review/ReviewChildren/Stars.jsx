import React from 'react'
import Star from '../../../star.jsx';


const Stars = (props) => {
  /*stars variables*/
  let nOfRatings = 0;
  let starsAvg = 0;
  let showNum = 0;
  let keys, values

  if (props['ratings'] !== undefined && props['ratings'][1]) {
    nOfRatings = Object.values(props['ratings']).reduce((a, b) => parseInt(a) + parseInt(b));
    keys = Object.keys(props['ratings'])
    values = Object.values(props['ratings'])
    let keyTimesValue = [];
    console.log(props['ratings'])
    for (var i=0; i<keys.length; i++) {
      keyTimesValue.push(keys[i] * values[i])
    }
    const reducer = (a, b) => a + b;
    starsAvg = keyTimesValue.reduce(reducer) / nOfRatings;
    showNum = starsAvg.toFixed(1)
  } else {
    return null
  }


  /*recommendation logic*/
  let falseRecs = 0;
  let trueRecs = 0;
  let percentage = 0;
  if (props.recommend !== undefined) {
    falseRecs = parseInt(props.recommend['false'])
    trueRecs = parseInt(props.recommend['true'])
    percentage = Math.round(((trueRecs / (trueRecs + falseRecs)) * 100))
  }
  console.log(nOfRatings)
  console.log(props['ratings'])
  return (
    <div>
      <div id="review-star-container">
        <div className="star-text">{showNum || null}&nbsp;</div>
        <div className="review-stars"> <Star rating={starsAvg}/></div>
      </div>

      <div id="review-rec">
        {percentage}% of reviews recommend this product<br />
      </div>

      <div className="bars-breakdown">
        <div id="breakdown">
          <label htmlFor="breakdown-rating">5 stars&nbsp;</label>
          <progress id="breakdown-rating" max="100" value={(props['ratings'][5]/nOfRatings)*100}></progress>
        </div>
        <div id="breakdown">
          <label htmlFor="breakdown-rating">4 stars&nbsp;</label>
          <progress id="breakdown-rating" max="100" value={(props['ratings'][4]/nOfRatings)*100}></progress>
        </div>
        <div id="breakdown">
          <label htmlFor="breakdown-rating">3 stars&nbsp;</label>
          <progress id="breakdown-rating" max="100" value={(props['ratings'][3]/nOfRatings)*100}></progress>
        </div>
        <div id="breakdown">
          <label htmlFor="file">2 stars&nbsp;</label>
          <progress id="breakdown-rating" max="100" value={(props['ratings'][2]/nOfRatings)*100}></progress>
        </div>
        <div id="breakdown">
          <label htmlFor="breakdown-rating">1 star&nbsp;</label>
          <progress id="breakdown-rating" max="100" value={(props['ratings'][1]/nOfRatings)*100}></progress>
        </div>
      </div>
    </div>
  )
}


export default Stars