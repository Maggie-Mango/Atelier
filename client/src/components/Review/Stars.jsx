import React from 'react'


const Stars = (props) => {
  /*stars variables*/
  let nOfRatings = 0;
  let starsAvg = 0;
  let starsFill = 0;
  let showNum = 0;
  let keys, values
  /*star bars variables*/
  let keyValArr = []
  let starBars;

  if (props.ratings !== undefined) {
    nOfRatings = Object.values(props.ratings).reduce((a, b) => parseInt(a) + parseInt(b))
    keys = Object.keys(props.ratings)
    values = Object.values(props.ratings)
    let keyTimesValue = [];
    for (var i=0; i<keys.length; i++) {
      keyTimesValue.push(keys[i] * values[i])
      keyValArr.push(keys[i])
      keyValArr.push(values[i])
    }
    const reducer = (a, b) => a + b;
    starsAvg = keyTimesValue.reduce(reducer) / nOfRatings;
    starsFill = (2.5/5) * 100
    showNum = starsAvg.toFixed(1)
    /*starbars logic*/
    starBars = keyValArr.map((item, index) =>
      index%2 === 0 ?
        index === 0 ?
      <div className="starbar-rating" key={index}>{item} star</div> :
      <div className="starbar-rating" key={index}>{item} stars</div>
      :
      <div className="bar-container">
        <div className="starbar-bar" key={index} style={{"width": `${(item / 5)*100}%` }}><br/></div>
      </div>

    )

  }


  /*recommendation logic*/
  let falseRecs = 0;
  let trueRecs = 0;
  let percentage = 0;
  if (props.recommend !== undefined) {
    falseRecs = parseInt(props.recommend['false'])
    trueRecs = parseInt(props.recommend['true'])
    percentage = Math.round(((trueRecs / (trueRecs + falseRecs)) * 100))
  }

  return (
    <div>
      <div className="star-text">{showNum} &nbsp;</div>
      <div className="star-rating">
        <span className="off"></span>
        <span className="on" style={{"width": `${starsFill}%`}}></span>
      </div>
      <div id="review-rec">
        {percentage}% of reviews recommend this product<br />
      </div>
      <div className="star-bars">
        {starBars}
      </div>

    </div>
  )
}


export default Stars