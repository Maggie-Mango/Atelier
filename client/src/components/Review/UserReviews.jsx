import React from 'react'
import test from './test-review.js'
import ReviewCard from './ReviewCard.jsx'

const UserReviews = (props) => {
  let reviews = props.reviews
  if (props.reviews.length > 0) {
    if (props.dropdown === 'relevance') {
      const relevance = [].concat(reviews)
      .sort((a, b) => a.helpfulness < b.helpfulness && a.date < b.date ? 1 : -1)
      .map( (item, i) => {
        return (
          <>
          <ReviewCard
            rating={item.rating}
            summary={item.summary}
            recommend={item.recommend}
            response={item.response}
            body={item.body}
            date={item.date}
            reviewer={item.reviewer_name}
            helpfulness={item.helpfulness}
            photos={item.photos}
          />
          </>
        )
      })
      return (
        <div>{relevance}</div>
      )
    }

    if (props.dropdown === 'helpfulness') {
      const helpful = [].concat(reviews)
      .sort((a, b) => a.helpfulness < b.helpfulness ? 1 : -1)
      .map( (item, i) => {
      return  (
        <>
        <ReviewCard
            rating={item.rating}
            summary={item.summary}
            recommend={item.recommend}
            response={item.response}
            body={item.body}
            date={item.date}
            reviewer={item.reviewer_name}
            helpfulness={item.helpfulness}
            photos={item.photos}
          />
        </>
      )
    })
    return (
      <div>{helpful}</div>
    )
    }

    if (props.dropdown === 'newest') {
      const recent = [].concat(reviews)
      .sort((a, b) => a.date < b.date ? 1 : -1)
      .map( (item, i) => {
        return  (
          <>
          <ReviewCard
            rating={item.rating}
            summary={item.summary}
            recommend={item.recommend}
            response={item.response}
            body={item.body}
            date={item.date}
            reviewer={item.reviewer_name}
            helpfulness={item.helpfulness}
            photos={item.photos}
          />
          </>
        )
      })
    return (
      <div>{recent}</div>
    )
    }


  }
  return (
    null
  )
}
/*
{
            "review_id": 1029620,
            "rating": 3,
            "summary": "Consequuntur unde quod velit vel.",
            "recommend": true,
            "response": "\"Ut maiores explicabo accusamus dicta.\"",
            "body": "Quisquam dicta vel. Quia nostrum suscipit provident voluptas illo tempora saepe enim. Aut distinctio ut aliquam distinctio quis consequatur maiores in. Nemo sint nobis dicta blanditiis sed et. Ut et enim non recusandae a est.",
            "date": "2021-04-09T00:00:00.000Z",
            "reviewer_name": "Clinton93",
            "helpfulness": 29,
            "photos": [
                {
                    "id": 1959391,
                    "url": "https://images.unsplash.com/photo-1503449377594-32dd9ac4467c?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1651&q=80"
                }
            ]
        }
*/

export default UserReviews