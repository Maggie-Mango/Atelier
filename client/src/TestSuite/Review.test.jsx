import React from 'react';
import {
  render,
  screen,
  mount,
  cleanUp,
  fireEvent,
  waitForElement
} from '@testing-library/react';
import '@testing-library/jest-dom'
import { unmountComponentAtNode } from "react-dom"


//Stars
import Stars from '../components/Review/Stars.jsx'

test('renders recommendation percentage', () => {
  render(<Stars />);
  const recommend = screen.getByText(/recommend/i);
  expect(recommend).toHaveTextContent('%')
  //screen.debug();
});

//AddReviewButton

//Characteristics

//Modal

//MoreReviewButton

//Review

//ReviewCard

//SortReview

//UserReviews
import UserReviews from '../components/Review/UserReviews.jsx'
test('UserReviews should contain ReviewCard ', () => {
  const detailsPanel = mount(
    <ReviewCard />);
  expect(detailsPanel.contains('Review Card')).toBeTruthy();
});