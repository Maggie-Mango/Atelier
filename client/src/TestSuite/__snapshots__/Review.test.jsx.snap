// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Breakdown correctly 1`] = `null`;

exports[`renders recommendation percentage 1`] = `null`;
